### PR TITLE
estimate number of PCs before embedding

### PR DESCRIPTION
**Pull request by gerbeldo**
# Background
#### Link to issue 
https://biomage.atlassian.net/browse/BIOMAGE-1387

#### Link to staging deployment URL 
https://ui-german-api268-pipeline193.scp-staging.biomage.net/

#### Links to any Pull Requests related to this
https://github.com/biomage-ltd/api/pull/268

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [x] Unit tests written
- [X] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [x] Deployed to staging
- [ ] Passed integration tests 

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [X] Photo of a cute animal attached to this PR
image

### CODE REVIEW

File: pipeline-runner/R/qc-6-integrate_scdata.R [hunk 1]
```diff
@@ -268,3 +287,12 @@ colorObject <- function(data) {
...
   return(data)
 }
+
+estimate_npcs <- function(scdata, var_threshold, max_npcs) {
+  # estimates the number of PCs to use in data integration and embeddings,
+  # using accumulated explained variance
+  eigValues <- (scdata@reductions$pca@stdev)^2
+  varExplained <- eigValues / sum(eigValues)
+  npcs <- min(which(cumsum(varExplained) >= var_threshold))
+  min(npcs, max_npcs, na.rm = TRUE)
```
Comment (alexvpickering) [inline pipeline-runner/R/qc-6-integrate_scdata.R:297]:
```suggestion
  eig_values <- (scdata@reductions$pca@stdev)^2
  var_explained <- eig_values / sum(eig_values)
  npcs <- min(which(cumsum(var_explained) >= var_threshold))
  min(npcs, max_npcs, na.rm = TRUE)
```

Comment (alexvpickering) [reply pipeline-runner/R/qc-6-integrate_scdata.R:297]: - The norm at least right now is snake case for variable names (functions also generally snake case in pipeline).

- What is the rationale for upper threshold on `npcs`?
- How did you decide on this method for choosing npcs? It doesn't seem to be among the methods suggested either in the [Bioconductor world](http://bioconductor.org/books/3.14/OSCA.advanced/dimensionality-reduction-redux.html#using-the-elbow-point) or [Seurat vignettes](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#determine-the-dimensionality-of-the-dataset)

File: pipeline-runner/R/qc-6-integrate_scdata.R [hunk 2]
```diff
@@ -235,6 +248,12 @@ run_unisample <- function(scdata, config) {
   scdata <- Seurat::ScaleData(scdata, verbose = FALSE)
   scdata <- Seurat::RunPCA(scdata, npcs = 50, features = Seurat::VariableFeatures(object = scdata), verbose = FALSE)
 
+  if (is.null(npcs)) {
+    message("\nEstimating number of PCs for UMAP\n")
+    npcs <- estimate_npcs(scdata, 0.85, 30)
+    message(sprintf("Number of PCs: %d\n", npcs))
```
Comment (alexvpickering) [inline pipeline-runner/R/qc-6-integrate_scdata.R:254]:
```suggestion
    message("Number of PCs: ", npcs)
```

`message` does a `paste0` and adds a newline (if you're going for double new line ignore - happy with whatever looks better in the terminal). e.g. try this:

```R
blah <- function() {
    npcs <- 30
    message("Number of PCs: ", npcs)
    message('hello')
}

blah()
```

File: pipeline-runner/R/qc-6-integrate_scdata.R [hunk 3]
```diff
@@ -268,3 +287,12 @@ colorObject <- function(data) {
   }
   return(data)
 }
+
+estimate_npcs <- function(scdata, var_threshold, max_npcs) {
```
Comment (alexvpickering) [inline pipeline-runner/R/qc-6-integrate_scdata.R:291]:
```suggestion
get_npcs <- function(scdata, var_threshold = 0.85, max_npcs = 30) {
```
just set as default parameter values and then remove arguments when calling function

File: pipeline-runner/R/qc-6-integrate_scdata.R [hunk 4]
```diff
@@ -26,6 +26,13 @@ integrate_scdata <- function(scdata, config, sample_id, cells_id, task_name = "d
   # main function
   set.seed(42)
   scdata.integrated <- run_dataIntegration(scdata, config)
+
+  if (is.null(config$dimensionalityReduction$numPCs)) {
```
Comment (alexvpickering) [inline pipeline-runner/R/qc-6-integrate_scdata.R:30]: do we need this check or can we just always run this code?

File: pipeline-runner/R/qc-6-integrate_scdata.R [hunk 5]
```diff
@@ -26,6 +26,13 @@ integrate_scdata <- function(scdata, config, sample_id, cells_id, task_name = "d
...
+    # get estimated npcs from the UMAP call in integration functions
+    estimated_npcs <- length(scdata.integrated@commands$RunUMAP@params$dims)
```
Comment (alexvpickering) [inline pipeline-runner/R/qc-6-integrate_scdata.R:32]:
```suggestion
    npcs <- length(scdata.integrated@commands$RunUMAP@params$dims)
```

File: pipeline-runner/R/qc-6-integrate_scdata.R (hunk 2 above)
Comment (gerbeldo) [reply pipeline-runner/R/qc-6-integrate_scdata.R:254]: neat. didn't know that. I added the newlines for _pretty printing_, but the messages might be useful in a development scenario, so we could probably do without most of them.

in any case, for the double newline we could use `message("\nNumber of PCs:", npcs, "\n")`

File: pipeline-runner/R/qc-6-integrate_scdata.R [hunk 6]
```diff
@@ -122,6 +128,11 @@ run_harmony <- function(scdata, config) {
   scdata <- add_dispersions(scdata)
   scdata@misc[["active.reduction"]] <- "harmony"
 
+  if (is.null(npcs)) {
+    npcs <- get_npcs(scdata)
+    message("Estimated number of PCs: ", npcs)
```
Comment (alexvpickering) [inline pipeline-runner/R/qc-6-integrate_scdata.R:133]:
```suggestion
    message("Number of PCs: ", npcs)
```

File: pipeline-runner/R/qc-6-integrate_scdata.R [hunk 7]
```diff
@@ -235,6 +246,11 @@ run_unisample <- function(scdata, config) {
   scdata <- Seurat::ScaleData(scdata, verbose = FALSE)
   scdata <- Seurat::RunPCA(scdata, npcs = 50, features = Seurat::VariableFeatures(object = scdata), verbose = FALSE)
 
+  if (is.null(npcs)) {
+    npcs <- get_npcs(scdata)
+    message("Estimated number of PCs: ", npcs)
```
Comment (alexvpickering) [inline pipeline-runner/R/qc-6-integrate_scdata.R:251]:
```suggestion
    message("Number of PCs: ", npcs)
```

**Review by alexvpickering — APPROVED** [review]
Check if this discussed likely bug is related to your PR. To reproduce:

- Run Vicky's WT1, WT2, and KO then change integration method to FastMNN

image

If it's related please fix first otherwise merge and create a ticket. Nice work!

**Pull request merged**